app.go

```diff
@@ -0,0 +1,423 @@
+		return
+	}
+
+	log.Infof("Fetching async task %s", id)
+
+	tx, err := a.db.BeginTx(context.TODO(), nil)
+	if err != nil {
+		errored(writer, err.Error())
+		return
+	}
+	defer tx.tx.Rollback()
+
+	task, err := tx.GetTask(id)
+	if err != nil {
+		errored(writer, err.Error())
+		return
+	}
+
+	if task.ID == "" {
+		notFound(writer, "not found")
+		return
+	}
+
+	err = tx.DeleteTask(id)
+	if err != nil {
+		errored(writer, err.Error())
+		return
+	}
+
+	tx.tx.Commit()
+	return
+}
+
+func (a *AsyncTasksApp) GetByFilterRequest(writer http.ResponseWriter, r *http.Request) {
+	var (
+		v = r.URL.Query()
+
+		filters = TaskFilter{
+			IDs:       v["id"],
+			Types:     v["type"],
+			Statuses:  v["status"],
+			Usernames: v["username"],
+		}
+		start_date_since  = v["start_date_since"]
+		start_date_before = v["start_date_before"]
+		end_date_since    = v["end_date_since"]
+		end_date_before   = v["end_date_before"]
+		null_end          = v["include_null_end"]
+	)
+
+	if len(null_end) > 0 {
+		filters.IncludeNullEnd = true
+	}
+
+	for _, startdate := range start_date_since {
+		parsed, err := time.Parse(time.RFC3339Nano, startdate)
+		if err != nil {
+			errored(writer, err.Error())
+			return
+		}
+		filters.StartDateSince = append(filters.StartDateSince, parsed)
+	}
+
+	for _, startdate := range start_date_before {
+		parsed, err := time.Parse(time.RFC3339Nano, startdate)
+		if err != nil {
+			errored(writer, err.Error())
+			return
+		}
+		filters.StartDateBefore = append(filters.StartDateBefore, parsed)
+	}
+
+	for _, enddate := range end_date_since {
+		parsed, err := time.Parse(time.RFC3339Nano, enddate)
+		if err != nil {
+			errored(writer, err.Error())
+			return
+		}
+		filters.EndDateSince = append(filters.EndDateSince, parsed)
+	}
+
+	for _, enddate := range end_date_before {
+		parsed, err := time.Parse(time.RFC3339Nano, enddate)
+		if err != nil {
+			errored(writer, err.Error())
+			return
+		}
+		filters.EndDateBefore = append(filters.EndDateBefore, parsed)
+	}
+
+	tx, err := a.db.BeginTx(context.TODO(), nil)
+	if err != nil {
+		errored(writer, err.Error())
+		return
+	}
+	defer tx.tx.Rollback()
+
+	tasks, err := tx.GetTasksByFilter(filters)
+
+	log.Info(tasks)
+
+	jsoned, err := json.Marshal(tasks)
+	if err != nil {
+		errored(writer, err.Error())
+		return
+	}
+
+	writer.Write(jsoned)
+
+	return
+}
+
+func (a *AsyncTasksApp) CreateTaskRequest(writer http.ResponseWriter, r *http.Request) {
+	var rawtask AsyncTask
+
+	body, err := ioutil.ReadAll(io.LimitReader(r.Body, 104857600))
+	if err != nil {
+		errored(writer, err.Error())
+		return
+	}
+	if err := r.Body.Close(); err != nil {
+		errored(writer, err.Error())
+		return
+	}
+	if err := json.Unmarshal(body, &rawtask); err != nil {
+		errored(writer, err.Error())
+		return
+	}
+
+	if rawtask.Type == "" {
+		errored(writer, "Task type must be provided")
+		return
+	}
+
+	for _, behavior := range rawtask.Behaviors {
+		if behavior.BehaviorType == "" {
```

Comment:
Some of this boilerplate makes me appreciate (even more) the auto type-conversions and error-checking `compojure-api` provides for us in our other RESTful services.

Reply:
Yeah, indeed. I might look into `echo` that Sarah (I think) found for the stats service work, because that might do some of this? I'm not sure though.

Reply:
Hm, echo doesn't provide validation directly, though it does provide a hook to use a third-party validation library I guess: https://echo.labstack.com/guide/request#validate-data -- it does seem to do some binding to structs some sort of automatically, which I need to read more about to see how it works.

Likely it's worth trying to integrate it anyway though. It looks pretty nice and it's a way to get swagger docs the way we like (similar to compojure-api and such).